### ISSUE TITLE
[FEATURE-REQUEST] New placeholder `ALL_CHANGED_FILES`

### ISSUE THREAD

**Issue by Eydamos** [captainhook-git/captainhook#235]
I have run into another edge case. For me it is common to amend commits and force push the changes to the remote.
If I use the `CHANGED_FILES` placeholder it will only list the files listed in the diff between the local hash and the remote hash. As git does not tell the hook that it is a force push there is no way around this behaviour.

Therefore I would propose the following solution:
If you would add a `ALL_CHANGED_FILES` placeholder that always uses the hash from the reflog as the `$oldHash` for comparison (just like you now do if the oldHash is just zeros) it would give me always a list of all files that have changed in the current branch compared to its original branch.

For my use-case I basically only want to make sure that when you push, one of the files that was changed between the current branch and the branch it was created from is the CHANGELOG.md

**Comment by sebastianfeldmann** [comment]
That's strange. Force pushing works fine for me.

- Change file X
- Commit
- Push the change (CHANGED_FILES => X)
- Amend commit 
- Force push (CHANGED_FILES => X)

Git puts the following into the `stdIn`

    refs/heads/main a74336878c058b6902df202db2f2253868cf2421 refs/heads/main 6e1fcae8e07c60775a9d777cbcfe54b3c63566d8

 `a74336` is the new commit hash `6e1fcae` is the amended old hash.

The detected revision range `6e1fcae.. a74336` works just fine to detect the files changed made with the amend.

Could you list a more detailed list of commands you execute to reproduce the problem?

**Comment by sebastianfeldmann** [comment]
And I would think that's correct :)

Because `CHANGELOG.md` is already on the remote in exactly that state.
So if you have checks on `.md` files it should have been checked before.

If you use changes to `.md` files to trigger some other local workflow I would suggest writing a custom `Condition` or `Action` to implement the behavior you need.

**Comment by Eydamos** [comment]
Of course this is correct. Which is why I requested a `ALL_CHANGED_FILES` placeholder which shows all changed files since the creation of the branch. I at least see a use case for this and unfortunately there is also no way how I could add custom placeholders.

If you think this is not useful this is totally fine. Then I will just write a whole custom action which get's the changes from the reflog. As I have both the repository and IO it should be possible.

I just want to make sure that in your current branch you have modified the CHANGELOG.md file as sometimes people forget this and sometimes you also do not realised this in the PR and merge it without the CHANGELOG being changed
```
{
    "pre-push": {
        "enabled": true,
        "actions": [
            {
                "action": "echo {$ALL_CHANGED_FILES|of-type:md} | grep 'CHANGELOG.md'",
                "config": {
                    "label": "Check for 'CHANGELOG.md'"
                },
                "conditions": [
                    {
                        "exec": "\\Company\\Project\\Condition\\NotOnBranch",
                        "args": [
                            "master"
                        ]
                    }
                ]
            }
        ]
    }
}
```

**Comment by sebastianfeldmann** [comment]
I think this is what you need

    git log --name-only --format='' main..HEAD | grep CHANGELOG.md

If you are sure that the main branch is the origin of your branch you can use this.

**Comment by Eydamos** [comment]
I can not just compare to master. We have dozens of repositories so I made one repository that has a captainhook config with all the hooks. Some of those repositories have `master` some have `main` and some `2.x` so I can not hardcode this.

BTW beside the `NotOnBranch` condition I also have a `NotOnBranchByRegEx` so I can compare it against `/\\d.x/` ;)

**Comment by sebastianfeldmann** [comment]
Mh playing around with `5.20.*` I run into more an more problems with the `reflog` solution.

If I push and then tag the repo and then just push the repo I get a zero hash for the tag because it does not exist on the remote.

This leads the Cap'n to use the `reflog` solution to figure out the changes.
While working on the main branch this now returns basically all files of the repository :)
Not sure that makes sense

I would have to check if the rev I'm receiving via `stdIn` is a tag or a commit.
Maybe it would be best to introduce a new placeholder for the `reflog` stuff because now it causes more problems than it solves.

**Comment by sebastianfeldmann** [comment]
So here are my thoughts:

 - So far the `reflog` solution seems pretty unstable especially when working on `main` directly.
 - I will most likely deactivate the `reflog` fallback and return to the `CHANGED_FILES` behavior < `5.20.0`.
 - I will introduce a new placeholder and condition to check the changed files for the current branch.
 - I'm struggling to find a good name for the placeholder, and I'm still not sure that `reflog` is the best course of action.

#### Ideas so far

`{$BRANCH_CHANGES}`, `{$BRANCH_CHANGES|compared-to:master}`

```
                "conditions": [
                    {
                        "exec": "\\Company\\Project\\Condition\\BranchChanges",
                        "args": [
                            "master"
                        ]
                    }
                ]
 ```

If the source branch is not specified I could try to find it using the `reflog`.
What do you think?

**Comment by Eydamos** [comment]
I would go with option 2 `{$BRANCH_CHANGES}` as it is basically what I initially proposed and fits our needs and work style the most.
I don't mind checking all the files on every push it is actually what I really want

**Comment by Eydamos** [comment]
BTW the issue you had with the tags is exactly why I have the condition `NotOnBranch`. Tags are only pushed from the master branch so I skip the check for this branch as the branch itself is protected any way so nobody could push actual changes to it

**Comment by sebastianfeldmann** [comment]
Ok, then I will rollback the `{$CHANGED_FILES}` behavior and introduce the new `{$BRANCH_CHANGES}` placeholder and `Condition`.

**Comment by Eydamos** [comment]
I get an exception:
```
Command failed:
  exit-code: 128
  message:   fatal: ambiguous argument '73cf48c..HEAD..': unknown revision or path not in the working tree.
```
The issue comes from calling `Operator\Log::getChangedFilesSince()` as `$revision` you give it `$start . '..HEAD'` but `Log\Log::byRange()` expects a `$from` and a `$to` param instead of a single string

**Comment by sebastianfeldmann** [comment]
ARGHHH, will fix in a sec stay tuned :)

**Comment by sebastianfeldmann** [comment]
Should be fixed in `5.21.2` released 30 sec ago :)
